packages/atomic/src/utils/initialization-utils.tsx

```diff
@@ -218,15 +202,19 @@ export function BindStateToController(
         );
       }
 
-      unsubscribe = this[controllerProperty].subscribe(() => {
+      unsubscribeController = this[controllerProperty].subscribe(() => {
         this[stateProperty] = this[controllerProperty].state;
         options?.onUpdateCallbackMethod &&
           this[options.onUpdateCallbackMethod]();
       });
     };
 
+    const disconnectUnsubscribeDelay = 100;
     component.disconnectedCallback = function () {
-      unsubscribe();
+      setTimeout(() => {
```

Comment:
a setTimeout with an arbitrary value does the job well here, but I'm open to other methods

Reply:
You said in the description that sometimes the resubscribe was not happening correctly. Was it because Stencil was not calling `connectedCallback` in these cases? Or on the headless side?

If Stencil, maybe we could we report the bug to the ionic team?

Reply:
Stencil was calling connected callback correctly. The issue is more a misuse on my part, of doing the same thing on both connectedCallback/disconnectedCallback: assigning the `unsubscribe` value

The custom element instance and its properties remain the same after its created.

When moving the component in the DOM, you trigger a disconnectedCallback & a connectedCallback, those are from the custom element API in the browser, so probably not Stencil specific, which builds native custom elements.

I *think* the logical cause from this type of hard-to-explain intermittent issue is a race condition.

On connectedCallback:
```
unsubscribe = this[controllerProperty].subscribe
```

On disconnectedCallback
```
unsubscribe();
```

If disconnectedCallback is called before, everything works as expected, if connectedCallback is called before, you might see the component not being updated when the controller updates.

The amount of times facets are moved in the DOM, unsubscribing and resubscribing to it was not the best solution as well.

Reply:
* Could we mitigate the problem by changing how we move facets in facet manager ? Is there a method that would cause them to not become disconnected from the dom ? 🤔  Is that even possible ?

* Reading on MDN: 
https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements#using_the_lifecycle_callbacks

> connectedCallback: Invoked each time the custom element is appended into a document-connected element. This will happen each time the node is moved, and may happen before the element's contents have been fully parsed.
Note: connectedCallback may be called once your element is no longer connected, use Node.isConnected to make sure.

Is that something we could leverage ? The `Node.isConnected` property ?

Instead of an arbitrary delay, we use `attributeChangedCallback` + observe `isConnected` ?

Reply:
> Could we mitigate the problem by changing how we move facets in facet manager ? Is there a method that would cause them to not become disconnected from the dom ? 🤔 Is that even possible ?

No i've looked for that, either you move it in the DOM, or you change the order by css, which is an accessibility nono

> Is that something we could leverage ? The Node.isConnected property ?

It seems to be similar to what I do with `document.contains` we would still need the small delay to prevent race conditions of the disconnected/reconnected, my current fix makes it not necessary to resubscribe to the state, since the instance of the Facet component and the controller is always the same. I can take a look at isConnected, perhaps it removes the necessity of the set timeout

`attributeChangedCallback` is like the watch of a prop, I don't see the link here, can you expand?

>  Invoked each time one of the custom element's attributes is added, removed, or changed.

Reply:
🙌   Seems checking for `isConnected` when `disconnectedCallback` is called works well from my tests, good call @olamothe!! It wasn't the case with `document.contains`.